GoogleForCreators/web-stories-wp
# PPC Trigger: Publish w/ High Priority Items

## Context

If a user attempts to publish a story right now without a title, they get a modal suggesting they add a title. This is to update that suggestion to review everything that's high priority. 

## Summary

If a user has high priority items to review and attempts to publish they will get a new dialog and if they select to review the prepublish tab will become active and focused. 

## Relevant Technical Choices

Because of how the context layers are structured I had to jump through a few hoops, I will walk you through them here: 
1. When to show dialog to review checklist items: new state in `usePrepublishChecklist` called `isHighPriorityEmpty`. Given the current behavior of the title missing dialog I think it's smart to show this dialog until the issues are resolved so they have to acknowledge that there's things that could help make their story better. If `isHighPriorityEmpty` is true then the dialog is skipped over.
2. While `isHighPriorityEmpty` is false we return the review dialog. 
3. They can continue to publish as before with the missing title dialog 
4. OR they can review the checklist. 
5. If they click to review, we're using `focusChecklistTab` also from prepublish checklist provider. From `focusChecklistTab` in `prepublishChecklistProvider` we dispatch the `PPC_CHECKPOINT_ACTION` for `ON_PUBLISH_CLICKED` to update the currentCheckpoint to `ALL` and set `setFirstPublishAttempt` to true. 
6. `isFirstPublishAttempt` is getting watched in a `useEffect` hook in the inspector provider. When it becomes true and it's the first time it's true then we focus the prepublish checklist tab regardless of if it's active or not. This prevents the dialog from getting triggered again when the publish button is clicked after review (per discussion with design) 
7. If a story is swapped back to a draft this trigger's reset and the dialog will show again. 

The reason for these hoops is that the publish button can see the prepublish context but not the inspector, the inspector can see prepublish though. 
Looks like this: 
```
<PrepublishProvider>
<PublishButton />
<InspectorProvider />
</PrepublishProvider>
```

- On publish button (in Header) we only have access to the prepublish context, not the inspector. Add new callback to prepublish context that gets called on dialog in header. That callback updates state that is getting watched in the inspector layout context, when that changes and qualifications are met trigger focus, set ref so it doesn't keep happening. Same thing for the switch to draft button.

- Added new checkpoint state for `NO_ISSUES` which will check for new issues that surface from a formerly perfect story. This also means that the shared icon used in the panel tab and publish button get updated without additional checks. 

## To-do


## User-facing changes

- No more 'missing title' dialog. Instead you get the new checklist review dialog. 
- Dialog only shows one time **as long as you interact with the dialog** (if you click outside the dialog you'll get it again, standard behavior there). 
- If you 'review' the checklist from the dialog you get sent to the checklist tab and both high priority and recommended sections are enabled. 
- Dialog will trigger on publish click regardless of what checkpoint you've hit with the prepublish checklist so long as you have high priority items that aren't resolved. 


| minimal story | recommended trigger with lots o pages | 
| --- | --- |
| ![recommend review trigger](https://user-images.githubusercontent.com/10720454/116323267-7b334500-a772-11eb-8ade-43ce9ea2877e.gif) | ![recommended trigger with lots o pages](https://user-images.githubusercontent.com/10720454/116323218-66ef4800-a772-11eb-8d66-e38d3361fe23.gif)|


## Testing Instructions

- click 'publish' on an empty story, see that you get the review checklist dialog. If you click 'review', you'll get sent to the checklist tab with everything enabled. If you click continue to publish, you'll publish your story. 
- Now swap that back to a draft and hit publish again, should repeat the above. 
- Now have a new story and add at least 5 pages to it so that recommended items in the checklist are enabled, now try to publish the story - you should still get the dialog. 
- Ok now have a perfect story with no high priority checklist items, hit publish. no dialog should show up. 

### QA






This PR can be tested by following these steps:

1.

### UAT







This PR can be tested by following these steps:

1.

## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #7141  #6985
